pkg/formats/sniffer.go

```diff
@@ -25,72 +42,148 @@ func (fs *Sniffer) SniffReader(f io.ReadSeeker) (Format, error) {
 	fileScanner := bufio.NewScanner(f)
 	fileScanner.Split(bufio.ScanLines)
 
-	formatType := ""
-	formatEncoding := ""
-	formatVersion := ""
+	var format Format
 
+	ctx := fs.setSniffState(context.Background())
 	for fileScanner.Scan() {
-		if strings.Contains(fileScanner.Text(), `"bomFormat"`) && strings.Contains(fileScanner.Text(), `"CycloneDX"`) {
-			formatType = "application/vnd.cyclonedx"
-			formatEncoding = JSON
-		}
+		format = fs.sniff(ctx, fileScanner.Bytes())
 
-		if strings.Contains(fileScanner.Text(), `"specVersion"`) {
-			parts := strings.Split(fileScanner.Text(), ":")
-			if len(parts) == 2 {
-				ver := strings.TrimPrefix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSpace(parts[1]), ","), "\""), "\"")
-				if ver != "" {
-					formatVersion = ver
-					formatEncoding = JSON
-				}
-			}
+		if format != EmptyFormat {
+			break
 		}
+	}
 
-		if strings.Contains(fileScanner.Text(), "SPDXVersion:") {
-			formatType = "text/spdx"
-			formatEncoding = "text"
+	if format != EmptyFormat {
+		return format, nil
+	}
 
-			for _, ver := range []string{"2.2", "2.3"} {
-				if strings.Contains(fileScanner.Text(), fmt.Sprintf("SPDX-%s", ver)) {
-					formatVersion = ver
-					break
-				}
-			}
-			break
+	fmt.Fprintf(
+		os.Stderr, "format: %s version: %s encodingq: %s\n",
+		format.Type(), format.Version(), format.Encoding(),
+	)
+
+	// TODO(puerco): Implement a light parser in case the string hacks don't work
+	return "", fmt.Errorf("unknown SBOM format")
+}
+
+func (fs *Sniffer) setSniffState(ctx context.Context) context.Context {
+	state := make(map[string]sniffState, len(sniffFormats))
+	return setSniffState(ctx, state)
+}
+
+func (fs *Sniffer) sniff(ctx context.Context, data []byte) Format {
+
+	for _, sniffer := range sniffFormats {
+		format := sniffer.sniff(ctx, data)
+		if format != EmptyFormat {
+			return format
 		}
+	}
+	return EmptyFormat
+}
+
+type sniffState struct {
+	Type     string
+	Version  string
+	Encoding string
+}
+
+func (st *sniffState) Format() Format {
+	if st.Type != "" && st.Encoding != "" && st.Version != "" {
+		return Format(fmt.Sprintf("%s+%s;version=%s", st.Type, st.Encoding, st.Version))
+	}
+
+	return EmptyFormat
+}
+
+type cdxSniff struct{}
 
-		// In JSON, the SPDX version field would be quoted
-		if strings.Contains(fileScanner.Text(), "\"spdxVersion\"") ||
-			strings.Contains(fileScanner.Text(), "'spdxVersion'") {
-			formatType = "text/spdx"
-			formatEncoding = JSON
-			if formatVersion != "" {
-				break
+func (c cdxSniff) Type() string {
+	return "cdx"
```

Comment:
We should use the constants here,  but is this method needed?
```suggestion
	return CDXFORMAT
```